### PR TITLE
[6.4] Fix `--event-stream-output-path` truncation with multiple test products

### LINKED ISSUE #10336: `--event-stream-output-path` is truncated by each test product when using the `swiftbuild` build system
### Is it reproducible with SwiftPM command-line tools: `swift build`, `swift test`, `swift package` etc?

- [x] Confirmed reproduction steps with SwiftPM CLI. The description text _must_ include reproduction steps with either of command-line SwiftPM commands, `swift build`, `swift test`, `swift package` etc.

### Description

When a package has more than one test target and `swift test` is run with
`--event-stream-output-path`, the resulting file contains only the events from the last
test product to run. Events from all earlier products are lost, with no diagnostic.

This appears to be an interaction between the `swiftbuild` build system and Swift
Testing's event stream writer, both of which are individually behaving as written:

1. `swift test` now defaults to `--build-system swiftbuild`
   (`BuildOptions._buildSystem` in `Sources/CoreCommands/Options.swift`).
2. `SwiftBuildSystem.builtTestProducts` returns one `BuiltTestProduct` per test product,
   whereas the `native` backend returns a single umbrella test bundle
   (`LLBuildDescription` passes `umbrellaProductName: nil`).
3. `TestRunner.testPerProduct` spawns one process per product and forwards the full
   command line to each, so every product receives the same
   `--event-stream-output-path` value.
4. Swift Testing opens that path with `FileHandle(forWritingAtPath:)`, which resolves to
   fopen mode `"wb"` — truncate on open. There is no append mode on that path
   (`Sources/Testing/ABI/EntryPoints/EntryPoint.swift`, event stream output block).

So N test products produce N truncating opens of a single path, and only the final
writer's output survives.

Filing here rather than against swift-testing because the defect looks like SwiftPM
handing a single-writer output path to N concurrent-by-design processes; swift-testing's
truncating open is reasonable for a path it expects to own exclusively. Happy to move it
if that read is wrong.

Symbol names rather than line numbers are used above since this was diagnosed against
`main`; the line numbers will drift.

### Expected behavior

The file at `--event-stream-output-path` contains the complete JSON Lines event stream for
the entire `swift test` invocation, covering every test product that ran.

### Actual behavior

The file contains only the events from the last test product executed. Test records and
events for all other products are absent. Exit status and console output are unaffected and
still report the full run, so the loss is silent.

Consumers of the stream — CI gates, reporters, coverage of which tests ran — therefore see
a partial run as a complete one.

### Steps to reproduce

1. Create a package with two test targets:

```swift
// swift-tools-version: 6.0
import PackageDescription

let package = Package(
    name: "EventStreamRepro",
    targets: [
        .target(name: "Core"),
        .testTarget(name: "ATests", dependencies: ["Core"]),
        .testTarget(name: "BTests", dependencies: ["Core"]),
    ]
)
```

```swift
// Sources/Core/Core.swift
public let value = 1
```

```swift
// Tests/ATests/ATests.swift
import Testing
@Test func testA() { #expect(true) }
```

```swift
// Tests/BTests/BTests.swift
import Testing
@Test func testB() { #expect(true) }
```

2. Run the suite, writing the event stream to a file:

```
swift test --disable-xctest \
  --event-stream-version 0 \
  --event-stream-output-path events.jsonl
```

3. List the modules present in the stream:

```
grep -o '"id":"[^".]*' events.jsonl | sort -u
```

**Observed:** only one of `ATests` / `BTests` appears.
**Expected:** both appear.

4. Re-run with the previous build system and repeat step 3:

```
swift test --build-system native --disable-xctest \
  --event-stream-version 0 \
  --event-stream-output-path events.jsonl
```

Both modules are now present, confirming the build system as the trigger.

### Swift Package Manager version/commit hash

Swift 6.4.0-dev

### Swift & OS version (output of `swift --version ; uname -a`)

swift-driver version: 1.168.5 Apple Swift version 6.4 (swiftlang-6.4.0.27.1 clang-2100.3.27.1)
Target: arm64-apple-macosx27.0.0
Darwin Tiny.local 27.0.0 Darwin Kernel Version 27.0.0: Mon Jun 29 21:26:34 PDT 2026; root:xnu-13432.0.50.501.3~1/RELEASE_ARM64_T8142 arm64

### PR DESCRIPTION
**Pull request by plemarquand**
Under the swiftbuild build system, `swift test` produces one test product per test target and runs each in its own process, forwarding the same command line to each. Swift Testing opens `--event-stream-output-path` in truncating mode (`fopen(path, "wb")`), so with N products every writer but the last is silently overwritten and the final file contains only the last product's events.

Unify the merging with the existing approach for `--xunit-output` implemented in #10263: give each
product its own per-product event-stream path in a temp dir, then concatenate them into the caller-specified path.

Fixes #10336

### CODE REVIEW

**Comment by plemarquand** [comment]
@swift-ci test

File: Tests/CommandsTests/FileContentsMergerTests.swift [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        try EventStreamMerger.merge(sources: [a, b], into: dest, fileSystem: fs)
+
+        let merged: String = try fs.readFileContents(dest)
+        let lines = merged.split(separator: "\n", omittingEmptySubsequences: true).map(String.init)
+        #expect(lines.count == 4, "every record from both sources must survive; got:\n\(merged)")
+        #expect(lines[0].contains("ATests.testA"))
+        #expect(lines[1].contains("testStarted"))
+        #expect(lines[2].contains("BTests.testB"))
+        #expect(merged.contains("ATests.testA"))
+        #expect(merged.contains("BTests.testB"))
```
Comment (bkhouri) [inline Tests/CommandsTests/FileContentsMergerTests.swift:60]: suggestion: can we also add


```suggestion
        #expect(merged.contains("BTests.testB"))
        #expect(merged.contains(sampleA))
        #expect(merged.contains(sampleB))
```

File: Tests/CommandsTests/EventStreamMergerTests.swift [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        #expect(lines.count == 4, "every record from both sources must survive; got:\n\(merged)")
+        #expect(lines[0].contains("ATests.testA"))
+        #expect(lines[1].contains("testStarted"))
+        #expect(lines[2].contains("BTests.testB"))
+        #expect(merged.contains("ATests.testA"))
+        #expect(merged.contains("BTests.testB"))
+    }
+
+    @Test
+    func mergesSingleSourceUnchanged() throws {
```
Comment (bkhouri) [inline Tests/CommandsTests/EventStreamMergerTests.swift]: suggestion: can we convert this to parameterized tests to validate various combinations while verify the merged line count is as expected, and that the merged contents ends with `\n`?

File: Tests/CommandsTests/FileContentsMergerTests.swift [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        let dest = AbsolutePath("/merged.jsonl")
+
+        try EventStreamMerger.merge(sources: [], into: dest, fileSystem: fs)
+
+        let merged: String = try fs.readFileContents(dest)
+        #expect(merged.isEmpty)
+    }
+
+    @Test
+    func ignoresBlankLinesBetweenRecords() throws {
```
Comment (bkhouri) [inline Tests/CommandsTests/FileContentsMergerTests.swift:132]: suggestion: Looking at this in isolation, can we add a test that merged empty files?

File: Sources/Commands/EventStreamMerger.swift [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import Foundation
+
+/// Combines Swift Testing event stream files produced by individual test
+/// binaries into a single JSON Lines output file. Every record from every
+/// source is preserved, in source order.
+///
+/// Multiple Swift Testing binaries (one per test product) each write their own
+/// event stream; without merging, later invocations would truncate earlier ones
+/// via `fopen(path, "wb")`.
+enum EventStreamMerger {
```
Comment (bkhouri) [inline Sources/Commands/EventStreamMerger.swift]: suggestion: the name is misleading as it's not specific to EventStream.  Can we rename this to "FileContentsMerger"?

**Comment by plemarquand** [comment]
@swift-ci test

File: Sources/Commands/FileContentsMerger.swift [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+/// via `fopen(path, "wb")`.
+enum FileContentsMerger {
+    static func merge(
+        sources: [AbsolutePath],
+        into destination: AbsolutePath,
+        fileSystem: FileSystem = localFileSystem,
+    ) throws {
+        var lines: [String] = []
+        for source in sources {
+            guard fileSystem.exists(source) else { continue }
```
Comment (pennig) [inline Sources/Commands/FileContentsMerger.swift:31]: Should an unexpectedly missing file have some diagnostics here, or are we comfortable dropping them on the floor for this codepath?

Comment (bkhouri) [reply Sources/Commands/FileContentsMerger.swift:31]: Diagnostics should be actionable to informed the user how to address it.  Emitting a diagnostic here would not be actionable, potentially leaving the user frustrated/annoyed.

**Review by bkhouri — APPROVED** [review]

**Comment by bkhouri** [comment]
@swift-ci test windows
